src/NewRelic.Core/NewRelic.SystemInterfaces/PerformanceCounterProxy.cs

```diff
@@ -11,178 +11,179 @@
+        /// <returns>the instance name or NULL if one could not be identified.</returns>
+        /// <param name="categoryName"></param>
+        public string GetCurrentProcessInstanceNameForCategory(string categoryName)
+        {
+            var processName = _processStatic.GetCurrentProcess().ProcessName;
+            var pid = _processStatic.GetCurrentProcess().Id;
+
+            var result = GetInstanceNameForProcessAndCategory(categoryName, processName, pid);
+
+            return result;
+        }
+
+        public IPerformanceCounterProxy CreatePerformanceCounterProxy(string categoryName, string counterName, string instanceName)
+        {
+            if (string.IsNullOrWhiteSpace(instanceName)) throw new ArgumentException(nameof(instanceName));
+            if (string.IsNullOrWhiteSpace(counterName)) throw new ArgumentException(nameof(counterName));
+            if (string.IsNullOrWhiteSpace(categoryName)) throw new ArgumentException(nameof(categoryName));
+
+            return _createPerformanceCounter(categoryName, counterName, instanceName);
+        }
+
+        /// <summary>
+        /// The instance name of a performance counter is a combination of the process Name and an index id to 
+        /// uniquely identify it when multiple of the same process are running.  For example "TestApp #1" pid=82, "TestApp #2" pid=134
+        /// are two instances of the same executable, TestApp.  In order to determine if an instance matches a process,
+        /// an additional performance counter "Process ID" must be used which returns the PID of the process.
+        /// </summary>
+        /// <param name="perfCategoryName"></param>
+        /// <param name="processName"></param>
+        /// <param name="pid"></param>
+        /// <returns>The instance name that will be used to collect performance counter data for the process</returns>
+        private string GetInstanceNameForProcessAndCategory(string perfCategoryName, string processName, int pid)
+        {
+            var performanceCategory = _createPerformanceCounterCategory(perfCategoryName);
+
+            var instanceNames = performanceCategory.GetInstanceNames()
+                .Where(x => x.StartsWith(processName, StringComparison.OrdinalIgnoreCase));
+
+            foreach (var instanceName in instanceNames)
+            {
+                try
+                {
+                    using (var pcPid = _createPerformanceCounter(perfCategoryName, ProcessIdCounterName, instanceName))
+                    {
+                        if ((int)pcPid.NextValue() == pid)
+                        {
+                            return instanceName;
+                        }
+                    }
+                }
+                catch (UnauthorizedAccessException)
+                {
+                    //The caller should handle this specificially.  It indicates that the
+                    //current user does not have access to pull performance counter values.
```

Comment:
Have you tried to validate this statement when dealing with 2 processes of the same app? Will one app get an unauthorized exception when trying to access the instance name of the other process?